zebra-chain/src/orchard_zsa/burn.rs

```diff
@@ -0,0 +1,80 @@
+//! Orchard ZSA burn related functionality.
+
+use std::io;
+
+use crate::{
+    amount::Amount,
+    block::MAX_BLOCK_BYTES,
+    serialization::{SerializationError, TrustedPreallocate, ZcashDeserialize, ZcashSerialize},
+};
+
+use orchard::note::AssetBase;
+
+use super::serialize::ASSET_BASE_SIZE;
+
+// Sizes of the serialized values for types in bytes (used for TrustedPreallocate impls)
+const AMOUNT_SIZE: u64 = 8;
+// FIXME: is this a correct way to calculate (simple sum of sizes of components)?
+const BURN_ITEM_SIZE: u64 = ASSET_BASE_SIZE + AMOUNT_SIZE;
+
+/// Represents an Orchard ZSA burn item.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct BurnItem(AssetBase, Amount);
+
+impl ZcashSerialize for BurnItem {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        let BurnItem(asset_base, amount) = self;
+
+        asset_base.zcash_serialize(&mut writer)?;
+        amount.zcash_serialize(&mut writer)?;
+
+        Ok(())
+    }
+}
+
+impl ZcashDeserialize for BurnItem {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        Ok(Self(
+            AssetBase::zcash_deserialize(&mut reader)?,
+            Amount::zcash_deserialize(&mut reader)?,
+        ))
+    }
+}
+
+impl TrustedPreallocate for BurnItem {
+    fn max_allocation() -> u64 {
+        // FIXME: is this a correct calculation way?
+        // The longest Vec<BurnItem> we receive from an honest peer must fit inside a valid block.
+        // Since encoding the length of the vec takes at least one byte, we use MAX_BLOCK_BYTES - 1
+        (MAX_BLOCK_BYTES - 1) / BURN_ITEM_SIZE
+    }
+}
+
+#[cfg(any(test, feature = "proptest-impl"))]
+impl serde::Serialize for BurnItem {
```

Comment:
> need to import `use serde::{Serialize, Deserialize};` here

There are already `Serialize` and `Deserialize` Serde derive macros in scope (added automatically via the `serde_derive` feature in `Cargo.toml`). If we were to override them by importing:

```rust
use serde::{Serialize, Deserialize};
```

it would work for this particular module since the derive macros are not used here. However, there are several places in the existing code of other modules where `serde::Serialize` is also implemented, and they do that using the full `serde::Serialize` name. This is because those modules do use the `Serialize` macro and cannot afford to overload it with use `serde::Serialize`. For example:

```rust
zebra-chain/src/orchard/shielded_data.rs: impl serde::Serialize for Flags { ... }
zebra-chain/src/orchard/tree.rs: impl serde::Serialize for Node { ... }
zebra-chain/src/sapling/tree.rs: impl serde::Serialize for Node { ... }
zebra-chain/src/sprout/tree.rs: impl serde::Serialize for Node { ... }
```

Therefore, I just followed the existing style used in all other parts of the code where `serde::Serialize` is implemented, but I can fix it for this module though.

Reply:
If I understand correctly, we do need `serde::Serialize` implemented for `BurnItem` because `burn` is now a part of the `OrchardShieldedData`:

https://github.com/QED-it/zebra/blob/zsa-integration-generics/zebra-chain/src/orchard/shielded_data.rs#L56

And this struct includes `Serialize` in its `#[derive(...)]`, even in the original Zebra code:

https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/orchard/shielded_data.rs#L24

We could simply add `Serialize` to the `#[derive(...)]` of `BurnItem` like this:

```rust
#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub struct BurnItem(AssetBase, Amount);
``

However, this approach doesn't work because `AssetBase` is imported from the `orchard` crate, and `serde::Serialize` is not implemented for it there. Therefore, I had to implement `serde::Serialize` manually for `BurnItem` in this module.